FWCore/Framework/src/EventProcessor.cc

```diff
@@ -536,9 +536,30 @@ namespace edm {
     actReg_->preallocateSignal_(bounds);
     schedule_->convertCurrentProcessAlias(processConfiguration_->processName());
     pathsAndConsumesOfModules_.initialize(schedule_.get(), preg());
+    std::set<ModuleProcessName> consumedBySubProcesses;
+    for_all(subProcesses_, [&consumedBySubProcesses](auto& subProcess) {
+      auto c = subProcess.deleteNonConsumedUnscheduledModules();
+      consumedBySubProcesses.insert(c.begin(), c.end());
+    });
 
-    //NOTE: this may throw
+    // Note: all these may throw
     checkForModuleDependencyCorrectness(pathsAndConsumesOfModules_, printDependencies_);
+    if (auto const unusedModules = nonConsumedUnscheduledModules(pathsAndConsumesOfModules_, consumedBySubProcesses);
+        not unusedModules.empty()) {
+      pathsAndConsumesOfModules_.removeModules(unusedModules);
+
+      edm::LogWarning("DeleteModules").log([&unusedModules](auto& l) {
+        l << "Following modules are not in any Path or EndPath, nor is their output consumed by any other module, and "
+             "therefore they are deleted before beginJob transition.";
+        for (auto const& description : unusedModules) {
+          l << "\n " << description->moduleLabel();
+        }
+      });
```

Comment:
I was thinking to keep this printout (and the corresponding one in the `SubProcess.cc`) enabled for a couple of IBs (to ease debugging in case problems occur), and then demote them to `LogInfo`.

Reply:
Sounds reasonable. In the future probably would want an easy way to make this list easily visible. Maybe via the Tracer?

Reply:
Tracer already prints out the deleted modules (between "ModuleConstruction" and "BeginJob"). Also with `LogInfo` the "DeleteModules" messages can be enabled purely in the configuration.